spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    CloseableIterable<IndexedTask> pendingTasks = getChanges(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId());
+    CloseableIterable<IndexedTask> rateLimitedTasks = rateLimit(pendingTasks, maxFilesPerTrigger);
+    IndexedTask last = Iterables.size(rateLimitedTasks) == 0 ? null : Iterables.getLast(rateLimitedTasks);
+
+    if (last == null) {
+      return start;
+    } else {
+      boolean isStarting = last.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(last.snapshotId(), last.index(), isStarting);
+    }
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> getChanges(long snapshotId, int index, boolean isStarting) {
+    List<CloseableIterable<IndexedTask>> indexedTasks = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+
+    if (isStarting) {
+      CloseableIterable<FileScanTask> iter = sortByFile(buildTableScan().useSnapshot(snapshotId).planFiles());
+      List<IndexedTask> tasks = Streams.mapWithIndex(Streams.stream(iter),
+          (t, i) -> new IndexedTask(snapshotId, t, (int) i))
+          .collect(Collectors.toList());
+      indexedTasks.add(CloseableIterable.combine(tasks, iter));
+    }
+
+    Long start = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(start != null, "Start snapshot id should exist");
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, start, currentSnapshotId))
+        .add(start)
+        .build()
+        .reverse();
+
+    for (int i = 0; i < snapshotIds.size() - 1; i++) {
+      long fromSnapshotId = snapshotIds.get(i);
+      long toSnapshotId = snapshotIds.get(i + 1);
+
+      CloseableIterable<FileScanTask> iter = sortByFile(
+          buildTableScan().appendsBetween(fromSnapshotId, toSnapshotId).planFiles());
+      List<IndexedTask> tasks = Streams.mapWithIndex(Streams.stream(iter),
+          (t, idx) -> new IndexedTask(toSnapshotId, t, (int) idx))
+          .collect(Collectors.toList());
+      indexedTasks.add(CloseableIterable.combine(tasks, iter));
+    }
+
+    return CloseableIterable.filter(CloseableIterable.concat(indexedTasks),
+        t -> t.snapshotId() != snapshotId || t.index > index);
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> rateLimit(CloseableIterable<IndexedTask> pendingTasks, int maxFiles) {
+    return CloseableIterable.combine(Iterables.limit(pendingTasks, maxFiles), pendingTasks);
+  }
+
+  private CloseableIterable<FileScanTask> sortByFile(CloseableIterable<FileScanTask> tasks) {
+    List<FileScanTask> sortedFiles = Streams.stream(tasks)
+        .sorted(COMPARATOR)
+        .collect(Collectors.toList());
```

Comment:
This is still loading all tasks into memory, which could be very large if the table is big and this needs to handle the whole table.

I think that sorting task files is probably the wrong design choice -- sorry that I suggested it! I think the right way is to disable the thread-pool used by `planFiles` so that the order is reliable. I suggested earlier to sort the manifest files, but disabling the thread-pool should have the same effect to make behavior deterministic.

I took a look at disabling the thread-pool and you can do it easily by adding an option to the scan, like this:

```java
  buildTableScan().option("use-worker-pool", "false").useSnapshotId(...).planFiles();
```

Then I changed `DataTableScan.planFiles` to use the new option:
```java
    boolean useWorkerPool = PropertyUtil.propertyAsBoolean(options(), "use-worker-pool", PLAN_SCANS_WITH_WORKER_POOL);
    if (useWorkerPool && snapshot.manifests().size() > 1) {
      manifestGroup.planWith(ThreadPools.getWorkerPool());
    }
```

Then you have deterministic behavior without loading all data files into memory at once.

Reply:
Updated.